runtime/src/lib.rs

```diff
@@ -411,6 +380,41 @@ impl pallet_usage::Config for Runtime {
 
 impl pallet_gas::Config for Runtime {}
 
+pub struct ExtraFeeFilter;
+impl Contains<Call> for ExtraFeeFilter {
+    fn contains(call: &Call) -> bool {
+        // Calls that affect message queue and are subject to extra fee
+        matches!(
+            call,
+            Call::Gear(pallet_gear::Call::submit_program { .. })
+                | Call::Gear(pallet_gear::Call::send_message { .. })
+                | Call::Gear(pallet_gear::Call::send_reply { .. })
+        )
+    }
+}
+
+type NegativeImbalance = <Balances as Currency<AccountId>>::NegativeImbalance;
+
+pub struct DealWithFees;
+impl OnUnbalanced<NegativeImbalance> for DealWithFees {
+    fn on_unbalanceds<B>(mut fees_then_tips: impl Iterator<Item = NegativeImbalance>) {
+        if let Some(fees) = fees_then_tips.next() {
+            if let Some(author) = Authorship::author() {
+                Balances::resolve_creating(&author, fees);
+            }
+            if let Some(tips) = fees_then_tips.next() {
+                if let Some(author) = Authorship::author() {
+                    Balances::resolve_creating(&author, tips);
+                }
```

Comment:
can't fee sum them in one operation?

Reply:
In Substrate `fee` and `tip` are treated slightly differently - could be the case for us, as well, so I kept it as is.
Anyway, I don't even remember if we have a mechanism for using tips at all - at least we don't create the signed extension explicitly with some tip value. So likely the second `if` is never triggered.

Most importantly, after we introduce staking this code will go. And this is going to happen real soon (hopefully).